extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,157 @@
+                    out.writeString(schemaJson);
+                    out.writeByteArray(datum instanceof LazyImmutableContainer
+                            ? ((LazyImmutableContainer<?>) datum).serialized
+                            : serialize(new GenericDatumWriter<>(datum.getSchema()), datum));
+                }
+
+                @Nonnull
+                @Override
+                public GenericContainer read(@Nonnull ObjectDataInput in) throws IOException {
+                    String schemaJson = in.readString();
+                    Schema schema = jsonToSchema.computeIfAbsent(schemaJson,
+                            json -> new Schema.Parser().parse(json));
+                    byte[] serializedDatum = in.readByteArray();
+                    switch (schema.getType()) {
+                        case RECORD:
+                            return new LazyImmutableRecord(serializedDatum, schema);
+                        case ARRAY:
+                            return new LazyImmutableArray<>(serializedDatum, schema);
+                        case ENUM:  // GenericEnumSymbol
+                        case FIXED: // GenericFixed
+                            return deserialize(new GenericDatumReader<>(schema), serializedDatum);
+                        default:
+                            throw new UnsupportedOperationException("Schema type " + schema.getType() + " is unsupported");
+                    }
+                }
+            };
+        }
 
         @Override
-        public Class<Utf8> getSerializationType() {
-            return Utf8.class;
+        public boolean isOverwritable() {
+            return false;
+        }
+    }
+
+    /**
+     * Replaces {@link Utf8} with {@link String} globally. Alternatively,
+     * the string class can be set on a per-schema basis via <ol>
+     * <li> {@link GenericData#setStringType GenericData.setStringType}{@code (schema,}
+     *      {@link StringType#String String}{@code )} or,
+     * <li> {@code SchemaBuilder.record("name").prop(}{@link GenericData#STRING_PROP STRING_PROP}{@code ,}
+     *      {@link StringType#String String}{@code )}.
+     */
+    public static final class CharSequenceHook implements SerializerHook<CharSequence> {
+        @Override
+        public Class<CharSequence> getSerializationType() {
+            return CharSequence.class;
         }
 
         @Override
         public Serializer createSerializer() {
-            return new StreamSerializer<Utf8>() {
+            return new ByteArraySerializer<CharSequence>() {
                 @Override
                 public int getTypeId() {
                     return SerializerHookConstants.AVRO_UTF8;
                 }
 
                 @Override
-                public void write(ObjectDataOutput out, Utf8 object) throws IOException {
-                    out.writeByteArray(object.getBytes());
+                public byte[] write(CharSequence cs) {
```

Comment:
`String` is also `CharSequence`. Will this hook be used instead of `StringSerializer`?

Reply:
It will override `StringSerializer` if `SerializationConfig#allowOverrideDefaultSerializers` is true, which is false by default. (You can see `AbstractSerializationService.serializerForClass()` for where the decision is made.) When it overrides, it will produce the same byte array as builtin `ObjectDataOutput.writeString()` implementations. So, the binary format may only differ if the user has custom `ObjectDataOutput` / `ObjectDataInput` or custom `Serializer` for `String`. Our serializer may not be registered at all if the user has already a custom `Serializer` for `CharSequence`. I implemented `Utf8` conversion this way by considering the rarity of such usage. However, as I indicated in the class Javadoc, there are more flexible (albeit tedious) solutions, which eliminate the conversion of `String` to `Utf8` in the first place.

Reply:
Is it the responsibility of the user to define STRING_PROP = String in their Map schemas? (I understand that way AVRO_COMPLEX_TYPES schema)

Reply:
does 👍 mean "yes"? If so, is that reasonable? Why Avro even has STRING_PROP in _schema_? Isn't that an implementation detail of ser/de, not something that affects the serialized binary form of data?

Reply:
but if the only consequence is that in order to user Avro MAP with string key users will have to define STRING_PROP for the MAP, then it is acceptable for now. Definitely much simpler than messing with CharSequence serialization.

Reply:
It is the user's responsibility to put `"avro.java.string": "String"` to _every_ schema for which they want to have `String` instead of `Utf8` on deserialization because this property is not inherited from parent schema.

I also think Avro shouldn't introduce a new class to represent UTF-8-encoded text, and Avro schemas shouldn't contain properties that are not about the serialized binary form of data. However, this is the case unfortunately, and the correct place to raise those objections is [Apache JIRA][1], where other people also complain about the inconveniences of `Utf8`.

[1]: https://issues.apache.org/jira/browse/AVRO-3848?jql=project%20%3D%20AVRO%20AND%20(summary%20~%20%22utf8%22%20OR%20summary%20~%20%22utf-8%22%20OR%20description%20~%20%22utf8%22%20OR%20description%20~%20%22utf-8%22)%20ORDER%20BY%20createdDate%20DESC